src/nvim/drawline.c

```diff
@@ -1661,15 +1662,36 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool nochange,
 
     if (wlv.draw_state == WL_LINE
         && has_fold
+        && !has_foldtext
         && wlv.col == win_col_offset
         && wlv.n_extra == 0
         && wlv.row == startrow + wlv.filler_lines) {
-      wlv.char_attr = win_hl_attr(wp, HLF_FL);
+      wlv.line_attr_lowprio = win_hl_attr(wp, HLF_FL);
+    }
 
+    if (wlv.draw_state == WL_LINE
+        && has_foldtext
+        && wlv.col == win_col_offset
+        && wlv.n_extra == 0
+        && wlv.row == startrow + wlv.filler_lines) {
+      wlv.char_attr = win_hl_attr(wp, HLF_FL);
       linenr_T lnume = lnum + foldinfo.fi_lines - 1;
       memset(buf_fold, ' ', FOLD_TEXT_LEN);
-      wlv.p_extra = get_foldtext(wp, lnum, lnume, foldinfo, buf_fold);
+      VirtText vt = KV_INITIAL_VALUE;
+      int vtwidth = 0;
+      wlv.p_extra = get_foldtext(wp, lnum, lnume, foldinfo, buf_fold, &vt, &vtwidth);
       wlv.n_extra = (int)strlen(wlv.p_extra);
+      if (vtwidth) {
+        for (int i = wlv.col; i < wlv.col + vtwidth; i++) {
+          linebuf_attr[i] = wlv.char_attr;
+        }
+
+        int col = draw_virt_text_item(buf, wlv.col, vt, kHlModeCombine, wlv.col + vtwidth, 0);
```

Comment:
This will not combine with line-wise highlights like 'cursorline' or sign "linehl".